flowapi/flowapi/permissions.py

```diff
@@ -0,0 +1,356 @@
+    q_treed : dict
+        Tree of queries with any references resolved
+
+
+    Returns
+    -------
+    dict
+        The new tree
+
+    Examples
+    --------
+    >>>build_tree(["DUMMY_ROOT"],{"DUMMY_ROOT": {"long_param": [1, 2, 3], "short_param": [1, 2]}})
+    {"DUMMY_ROOT": {"long_param": {1: {"short_param": {1: {}, 2: {}}},2: {"short_param": {1: {}, 2: {}}},3: {"short_param": {1: {}, 2: {}}},}}}
+
+    """
+    tree = {}
+    for root in roots:
+        params = q_treed[root]
+        tree[root] = {}
+        refs = tree[root]
+        sorted_params = sorted(params.items(), key=lambda x: (-len(x[1]), x[0]))
+        try:
+            longest_param, *others = sorted_params
+            param_name, param_spec = longest_param
+        except ValueError:
+            continue
+        tree[root][param_name] = {q_type: {} for q_type in param_spec}
+        refs = refs[param_name]
+        for param_name, param_spec in others:
+            for q_type, q_bod in refs.items():
+                refs[q_type][param_name] = {k: {} for k in param_spec}
+    return tree
+
+
+def enum_paths(parents: List[str], tree: dict) -> Tuple[List[str], dict]:
+    """
+    Yield the paths to the leaves of a tree and the associated leaf node.
+
+    Parameters
+    ----------
+    parents : list of str
+        Parents of this path
+    tree : dict
+        Tree of queries
+
+    Yields
+    ------
+    Tuple of list, dict
+    """
+    if len(tree) == 0:
+        yield parents, tree
+    else:
+        for k, v in tree.items():
+            yield from enum_paths(parents + [k], v)
+
+
+def make_per_query_scopes(tree: dict, all_queries: dict) -> Iterable[str]:
+    """
+    Constructs and yields query scopes of the form:
+    <query_kind>:<arg_name>:<arg_val>
+    where arg_val may be a query kind, or the name of an aggregation unit if applicable.
+
+    One scope is yielded for each viable query structure, so for queries which contain two child queries
+    two scopes are yielded. If that query has 3 possible aggregation units, then 6 scopes are yielded altogether.
+
+    Parameters
+    ----------
+    tree : dict
+        Dict of nested queries
+    all_queries : dict
+        All queries
+
+    Yields
+    ------
+    str
+        Query scope of the form <query_kind>:<arg_name>:<arg_val>
+
+    Examples
+    --------
+    >>>list(make_per_query_scopes({"DUMMY": {}}, {"DUMMY": {"query_kind": {"enum": ["dummy"]}}}))
+    ["dummy"]
+    >>>list(make_per_query_scopes({"DUMMY": {}},{"DUMMY": {"query_kind": {"enum": ["dummy"]},"aggregation_unit": {"enum": ["DUMMY_UNIT"]},}}))
+    ["dummy:aggregation_unit:DUMMY_UNIT",]
+
+    """
+    for path, _ in list(enum_paths([], tree)):
+        kind_path = [
+            all_queries.get(p, {}).get("query_kind", {}).get("enum", [p])[0]
+            for p in path
+        ]  # Want the snake-cased variant
+        try:
+            units = all_queries[path[-1]]["aggregation_unit"]["enum"]
```

Comment:
This will only create scopes for the aggregation units accepted by the last query kind in the path. If a top-level query takes two (or more) child queries as parameters then we may also need to check the aggregation units of query kinds further back in the path.

Let's suppose `exposed_query_kind` takes two parameters, one of which is a `child_query_kind_a` query and the other is `child_query_kind_b`. Then `path[-1]` will either be `"child_query_kind_a"` or `"child_query_kind_b"`, depending on the order in which the two parameters were added to the tree.

I think it would be excessive to allow aggregation unit permissions to be controlled separately for `child_query_kind_a` and `child_query_kind_b`, since the aggregation units for both would typically need to be the same (if both query kinds accepted an aggregation unit). But even with just one set of aggregation unit permissions  per path I don't think it's sufficient to check just the last element of the path.

E.g. if `path[-1]` is `child_query_kind_b`, which doesn't have an `aggregation_unit` parameter, but `child_query_kind_a` does, then `make_per_query_scopes` won't return scopes containing aggregation units even though `child_query_kind_a` has an `aggregation_unit` parameter that we might want to restrict permissions for. Or if `child_query_kind_a` accepts aggregation unit `agg1` or `agg2`, and `child_query_kind_b` accepts aggregation unit `agg1`, `agg2` or `agg3`, then `make_per_query_scopes` would return a scope with `aggregation_unit:agg3` even though `child_query_kind_a` doesn't accept this aggregation unit.